File: AccuPay/Repositories/DivisionRepository.vb
Option Strict On

Imports System.Threading.Tasks
Imports AccuPay.Entity
Imports Microsoft.EntityFrameworkCore

Namespace Global.AccuPay.Repository

    Public Class DivisionRepository

        Public Shared ReadOnly DIVISION_TYPE_DEPARTMENT As String = "Department"
        Public Shared ReadOnly DIVISION_TYPE_BRANCH As String = "Branch"
        Public Shared ReadOnly DIVISION_TYPE_SUB_BRANCH As String = "Sub branch"

        Public Function GetDivisionTypeList() As List(Of String)
            Return New List(Of String) From {
                        DIVISION_TYPE_DEPARTMENT,
                        DIVISION_TYPE_BRANCH,
                        DIVISION_TYPE_SUB_BRANCH
                }
        End Function

        Public Async Function GetAllAsync() As Task(Of List(Of Division))

            Using context As New PayrollContext

                Return Await context.Divisions.
                            Where(Function(d) Nullable.Equals(d.OrganizationID, z_OrganizationID)).
                            ToListAsync

            End Using

        End Function

        Public Async Function GetAllParentsAsync() As Task(Of List(Of Division))

            Using context As New PayrollContext

                Return Await context.Divisions.
                            Where(Function(d) Nullable.Equals(d.OrganizationID, z_OrganizationID)).
                            Where(Function(d) d.IsRoot).
                            ToListAsync

            End Using

        End Function

        Public Async Function SaveAsync(division As Division) As Task(Of Division)

            Using context As New PayrollContext

                Dim existingDivision As Division = Await GetByNameAndParentDivisionAsync(division.Name, division.ParentDivisionID)

                If division.RowID Is Nothing Then
                    Insert(division, existingDivision, context)

                Else
                    Update(division, existingDivision, context)
                End If

                Await context.SaveChangesAsync()


                Dim newDivision = Await context.Divisions.
                                    FirstOrDefaultAsync(Function(d) Nullable.Equals(d.RowID, division.RowID))

                If newDivision Is Nothing Then
                    Throw New ArgumentException("There was a problem inserting the new division location. Please try again.")
                End If

                Return newDivision
            End Using

        End Function

        Public Async Function DeleteAsync(divisionId As Integer?) As Task

            Using context = New PayrollContext()

                If context.AgencyFees.Any(Function(a) Nullable.Equals(a.DivisionID, divisionId)) Then

                    Throw New ArgumentException("Division already has agency fees therefore cannot be deleted.")

                ElseIf context.Divisions.Any(Function(d) Nullable.Equals(d.ParentDivisionID, divisionId)) Then

                    Throw New ArgumentException("Division already has child divisions therefore cannot be deleted.")

                ElseIf context.Positions.Any(Function(p) Nullable.Equals(p.DivisionID, divisionId)) Then

                    Throw New ArgumentException("Division already has positions therefore cannot be deleted.")

                End If

                Dim division = Await context.Divisions.FirstOrDefaultAsync(Function(d) Nullable.Equals(d.RowID, divisionId))

                context.Remove(division)

                Await context.SaveChangesAsync()

            End Using

        End Function

#Region "Private Function"

        Public Async Function GetByNameAndParentDivisionAsync(positionName As String, parentDivisionId As Integer?) As Task(Of Division)

            Using context As New PayrollContext

                Return Await context.Divisions.
                            Include(Function(d) d.ParentDivision).
                            Where(Function(d) d.Name.ToLower.Trim = positionName.ToLower.Trim).
                            Where(Function(d) Nullable.Equals(d.ParentDivisionID, parentDivisionId)).
                            Where(Function(d) Nullable.Equals(d.OrganizationID, z_OrganizationID)).
                            FirstOrDefaultAsync()

            End Using

        End Function

        Private Sub Insert(
                        division As Division,
                        existingDivision As Division,
                        context As PayrollContext)

            If existingDivision IsNot Nothing Then

                Throw New ArgumentException($"Division name already exists under { existingDivision.ParentDivision.Name }!")

            End If

            division.CreatedBy = z_User

            context.Divisions.Add(division)
        End Sub

        Private Sub Update(
                        division As Division,
                        existingDivision As Division,
                        context As PayrollContext)

            If existingDivision IsNot Nothing AndAlso
                Nullable.Equals(division.RowID, existingDivision.RowID) = False Then

                Throw New ArgumentException($"Division name already exists under { existingDivision.ParentDivision.Name }!")

            End If

            division.LastUpdBy = z_User

            context.Divisions.Attach(division)
            context.Entry(division).State = EntityState.Modified

        End Sub

#End Region

    End Class

End Namespace

File: Accupay.Data/Repositories/DivisionRepository.cs
using AccuPay.Data.Entities;
using AccuPay.Data.Helpers;
using AccuPay.Utilities.Extensions;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace AccuPay.Data.Repositories
{
    public class DivisionRepository : SavableRepository<Division>
    {
        public const string DIVISION_TYPE_DEPARTMENT = "Department";
        public const string DIVISION_TYPE_BRANCH = "Branch";
        public const string DIVISION_TYPE_SUB_BRANCH = "Sub branch";

        public DivisionRepository(PayrollContext context) : base(context)
        {
        }

        protected override void DetachNavigationProperties(Division entity)
        {
            if (entity.ParentDivision != null)
            {
                _context.Entry(entity.ParentDivision).State = EntityState.Detached;
            }
        }

        #region Queries

        #region Single entity

        public async Task<Division> GetByIdWithParentAsync(int id)
        {
            return await _context.Divisions
                .Include(x => x.ParentDivision)
                .FirstOrDefaultAsync(l => l.RowID == id);
        }

        public async Task<Division> GetOrCreateDefaultDivisionAsync(int organizationId, int userId)
        {
            using (var transaction = await _context.Database.BeginTransactionAsync())
            {
                try
                {
                    var defaultParentDivision = await _context.Divisions
                        .Where(x => x.OrganizationID == organizationId)
                        .Where(x => x.Name.Trim().ToLower() == Division.DefaultLocationName.ToTrimmedLowerCase())
                        .Where(x => x.IsRoot)
                        .FirstOrDefaultAsync();

                    if (defaultParentDivision == null)
                    {
                        defaultParentDivision = Division.NewDivision(
                            organizationId: organizationId,
                            userId: userId);

                        defaultParentDivision.Name = Division.DefaultLocationName;
                        defaultParentDivision.ParentDivisionID = null;
                        defaultParentDivision.LastUpdBy = userId;

                        //await SanitizeEntity(defaultParentDivision, null);
                        await SaveAsync(defaultParentDivision);
                        // querying the new default parent division from here can already
                        // get the new row data. This can replace the context.local in leaverepository
                    }
                    if (defaultParentDivision?.RowID == null)
                        throw new Exception("Cannot create default division location.");

                    var defaultDivision = await _context.Divisions
                        .Where(x => x.OrganizationID == organizationId)
                        .Where(x => x.Name.Trim().ToLower() == Division.DefaultDivisionName.ToTrimmedLowerCase())
                        .Where(x => x.ParentDivisionID == defaultParentDivision.RowID)
                        .FirstOrDefaultAsync();

                    if (defaultDivision == null)
                    {
                        defaultDivision = Division.NewDivision(
                            organizationId: organizationId,
                            userId: userId);

                        defaultDivision.Name = Division.DefaultDivisionName;
                        defaultDivision.ParentDivisionID = defaultParentDivision.RowID;
                        defaultDivision.LastUpdBy = userId;

                        //await SanitizeEntity(defaultDivision, null);
                        await SaveAsync(defaultDivision);
                    }

                    transaction.Commit();

                    return defaultDivision;
                }
                catch (Exception)
                {
                    transaction.Rollback();
                    throw;
                }
            }
        }

        #endregion Single entity

        #region List of entities

        public async Task<IEnumerable<Division>> GetAllAsync(int organizationId)
        {
            return await _context.Divisions
                .Where(d => d.OrganizationID == organizationId)
                .ToListAsync();
        }

        public async Task<PaginatedList<Division>> List(
            PageOptions options,
            int organizationId,
            string searchTerm = "")
        {
            var query = _context.Divisions
                .Include(x => x.ParentDivision)
                .Where(x => x.OrganizationID == organizationId)
                .OrderBy(x => x.ParentDivision.Name)
                .ThenBy(x => x.Name)
                .AsQueryable();

            if (!string.IsNullOrWhiteSpace(searchTerm))
            {
                searchTerm = $"%{searchTerm}%";

                query = query.Where(x =>
                    EF.Functions.Like(x.Name, searchTerm) ||
                    EF.Functions.Like(x.ParentDivision.Name, searchTerm));
            }

            var divisions = await query.Page(options).ToListAsync();
            var count = await query.CountAsync();

            return new PaginatedList<Division>(divisions, count);
        }

        public async Task<ICollection<Division>> GetAllParentsAsync(int organizationId)
        {
            return await _context.Divisions
                .Where(d => d.OrganizationID == organizationId)
                .Where(d => d.IsRoot)
                .ToListAsync();
        }

        #endregion List of entities

        #region Others

        public List<string> GetDivisionTypeList()
        {
            return new List<string>()
            {
                DIVISION_TYPE_DEPARTMENT,
                DIVISION_TYPE_BRANCH,
                DIVISION_TYPE_SUB_BRANCH
            };
        }

        #endregion Others

        #endregion Queries
    }
}
